From 33a49edbccc904e12dfc9ee1a43269457b92296e Mon Sep 17 00:00:00 2001
From: Colin Cross <ccross@android.com>
Date: Tue, 24 Jul 2012 13:27:57 -0700
Subject: [PATCH 610/944] power: smb347-charger: reset chip during init

Reset the charger chip during init.  It's going to reset it self during
normal operation every time the primary input is unplugged, so we might
as well reset it manually so it's always in a consistent state after
boot and after unplugging.

Change-Id: Ia50024ebb8fcd50eeef565ccea994076b3f04546
Signed-off-by: Colin Cross <ccross@android.com>
---
 drivers/power/smb347-charger.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/power/smb347-charger.c b/drivers/power/smb347-charger.c
index 8b76c1e..2f3ffb5 100644
--- a/drivers/power/smb347-charger.c
+++ b/drivers/power/smb347-charger.c
@@ -22,6 +22,7 @@
 #include <linux/power_supply.h>
 #include <linux/power/smb347-charger.h>
 #include <linux/seq_file.h>
+#include <linux/delay.h>
 
 /*
  * Configuration registers. These are mirrored to volatile RAM and can be
@@ -91,6 +92,7 @@
 #define CMD_A_OTG_ENABLE			BIT(4)
 #define CMD_A_ALLOW_WRITE			BIT(7)
 #define CMD_B					0x31
+#define CMD_B_POR				BIT(7)
 #define CMD_B_USB59_MODE			BIT(1)
 #define CMD_B_HC_MODE				BIT(0)
 #define CMD_C					0x33
@@ -1401,6 +1403,18 @@ static int smb347_probe(struct i2c_client *client,
 			smb->en_gpio = pdata->en_gpio;
 	}
 
+	ret = smb347_write(smb, CMD_B, CMD_B_POR);
+	if (ret < 0)
+		return ret;
+
+	msleep(20);
+
+	ret = smb347_read(smb, CMD_B);
+	if (ret < 0) {
+		dev_err(dev, "failed read after reset\n");
+		return ret;
+	}
+
 	ret = smb347_hw_init(smb);
 	if (ret < 0)
 		return ret;
-- 
1.8.0

